WordPress to Jetpack flow: main flow implementation

Gio2018 opened the pull request:
Ref #19517 
Ref #19518 

This PR implements the remaining wp-to-jp flow

NOTE: the UI is not finalized (text and images can be missing or wrong in the notifications and all done screens), nor optimized for landscape and multiple screens, the notifications permissions is not there yet, and the migration flow is shown by default at each launch. All these points will be addressed in separate PRs

To test:
***Scenario A - Notifications permissions undetermined (user has not made a decision)***
**Case 1: skip notifications permission decision:**
1. checkout this branch and build/run the Jetpack target
2. In the welcome screen, tap Continue and make sure the Notifications permissions screen appears
3. In the Notifications permissions screen, tap decide later and make sure the all done screen appears. 
4. In the all done screen, tap Finish and make sure the migration UI gets dismissed and the app UI appears.

**Case 2: skip notifications permission decision:**
same as above, except on step 3, tap "Continue" and make sure the notifications permission alert shows up, then:
- Make a decision
- Regardless the decision, make sure the all done screen appears, and continue from step 4 above

***Scenario B - Notifications permissions determined (user has already made a decision)***
Same as Scenario A - Case 1, except step 3 is missing: after tapping Continue in the welcome screen, make sure the all done screen appears. After that, continue normally with step 4

- Optional: do a smoke test and make sure the app behaves correctly.

## Regression Notes
1. Potential unintended areas of impact
None other the migration itself

2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] ~~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~~ Not released

Commit:
Add MigrationDpendencyContainer, dependency injection container for wp-to-jp migration

Commit:
Add MigrationNavigationController, custom, state-based navigation controller for wp-to-jp migration

Commit:
Update MigrationFlowCoordinator, extract ui logic from coordinator

Commit:
Add MigrationStepConfiguration, a refactor of existing configuration types

Commit:
Add MigrationStepView, a common base view for migration step screens

Commit:
Add MigrationNotificationsViewController, view controller for the wp-to-jp migration-notifications permissions step

Commit:
Add MigrationNotificationsViewModel, view model for the wp-to-jp migration-notifications permissions step

Commit:
Add MigrationDoneViewController, view controller for the wp-to-jp migration-final step

Commit:
Add MigrationDoneViewModel, view model for the wp-to-jp migration-final step

Commit:
Update WindowManager, default optional completion to nil

Commit:
Update JetpackWindowManager, refactor to enable showing the wp-to-jp migration flow, if needed

Commit:
Update MigrationStep, add missing steps and extract in separate file

Commit:
Update MigrationActionsView, refactors based on the refactor of configuration types

Commit:
Update MigrationHeaderView, refactors based on the refactor of configuration types

Commit:
Files and groups re-arranged in WordPress-to-Jetpack migration group

Commit:
Update MigrationStepConfiguration, put separate types in separate files

Commit:
Update MigrationDoneViewController, MigrationNotificationsViewController, add some comments

Commit:
Update MigrationDependencyContainer, remove incorrect comment

Commit:
Update MigrationStepView, remove unnecessary settings

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Common/MigrationDependencyContainer.swift:
```diff
@@ -0,0 +1,46 @@
+
+    func makeInitialViewController() -> UIViewController {
+        MigrationNavigationController(coordinator: migrationCoordinator,
+                                      migrationStack: [.welcome: makeWelcomeViewController(),
```
I'm not opposed to this approach of having a dictionary mapping `MigrationStep` to a `UIViewController` but I just want to point out that these view controllers will stay in memory even when they're dismissed ( not visible ).

Perhaps we can optimize this by injecting a closure like this:

```swift
let closure = { [weak self] (step: MigrationStep) -> UIViewController? in 
     switch step {
           case .welcome: return self?.makeWelcomeViewController()
           ...
     }
}
```

Gio2018 replied:
I like this idea: I added a factory that does exactly that, and we don't retain any unused vc in memory!

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Common/Navigation/MigrationFlowCoordinator.swift:
```diff
@@ -0,0 +1,26 @@
+import Combine
+
+/// Coordinator for the migration to jetpack flow
+final class MigrationFlowCoordinator: ObservableObject {
```
My understanding of `Coordinator` objects in the `MVVM-C` architecture is that they can instantiate and work with `UIViewController` objects ( e.g. this [article](https://medium.com/p/3960ad9a6d85) and [this](https://www.linkedin.com/pulse/understanding-mvvm-c-budhabhooshan-patil?trk=articles_directory) ). That's why in the previous implementation, the coordinator was responsible for mapping a `MigrationStep` to a `UIViewController` and also pushing the view controllers to the navigation stack.

But I also like the idea of a UI-agnostic coordinator. But in this case, this object seems to me more of a `ViewModel` than a `Coordinator`. Maybe we should rename it to `MigrationFlowViewModel`?

What do you think?

Gio2018 replied:
Thanks for the thoughtful design review and the following discussion we had on Slack! As discussed, we'll leave the current naming for now and re-consider it together with Tanner (avoid direct pinging on purpose 😄 ) when he comes back from the meetup!

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Welcome/MigrationWelcomeViewModel.swift:
```diff
@@ -0,0 +1,37 @@
+        },
+                                                                                      secondaryHandler: {
+
+        }))
```
This is just a nice-to-have, but I honestly find this code a bit hard to read 🙈

Maybe we should breakdown this complex expression into multiple variables, something like this:

```swift
let headerConfig = MigrationHeaderConfiguration(step: .welcome, multiSite: blogListDataSource.visibleBlogsCount > 1)
let actionsConfig = MigrationActionsViewConfiguration(step: .welcome, primaryHandler:...)
configuration = MigrationStepConfiguration(headerConfiguration: headerConfig, actionsConfig: actionsConfig)

```

Gio2018 replied:
Totally agree. I did this in all view models

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Welcome/MigrationWelcomeViewController.swift:
```diff
@@ -10,14 +10,18 @@ final class MigrationWelcomeViewController: UIViewController {
 
-    private let bottomSheet = MigrationActionsView()
+    private lazy var bottomSheet: MigrationActionsView = {
+        let actionsView = MigrationActionsView(configuration: viewModel.configuration.actionsConfiguration)
```
Regarding the `MigrationActionsView` class, I just want to mention that I was planning to rename it to something like `ActionsSheetView`, so that the component can be re-used in other screens. 

We already have a similar component in the Jetpack / WordPress Prologue Screen, and it's very possible that we might see this component again in another screen.

But right now, this component is more dependent on the `Migration` feature because the `MigrationActionsViewConfiguration`'s sole initializer takes a `MigrationStep` param.

It would be great to have an initializer that doesn't depend on `MigrationStep` and another "convenience" init that can depend on it. In other words, it would be great to have these 2 initializers:

<img width="400" alt="CleanShot 2022-11-07 at 16 37 56@2x" src="https://user-images.githubusercontent.com/9609223/200351159-168e2a6d-d8ab-49bd-91e5-4a32f3881dbf.png">

We can achieve the above by moving the `init(step: MigrationStep...)` init to an extension:

```swift
// This struct can be defined in the same file as `MigrationActionsView.swift` file or even embedded in the `MigrationActionsView` class.
struct MigrationActionsViewConfiguration {
    let primaryTitle: String?
    let secondaryTitle: String?
    let primaryHandler: (() -> Void)?
    let secondaryHandler: (() -> Void)?
}

// Have this extension in another file.
extension MigrationActionsViewConfiguration {

    init(step: MigrationStep, primaryHandler: (() -> Void)? = nil, secondaryHandler: (() -> Void)? = nil) {
        self.primaryHandler = primaryHandler
        self.secondaryHandler = secondaryHandler
        self.primaryTitle = Appearance.primaryTitle(for: step)
        self.secondaryTitle = Appearance.secondaryTitle(for: step)
    }

    private enum Appearance {
        // ...
    }
}
```

Would look like to hear your thoughts on making this component reusable in other screens other than `Migration` screens.

_P.S: This is definitely not urgent and we can address it later_

Gio2018 replied:
This makes total sense. I'd think that, if we want to make this view reusable, we should then move it to `WordPressUI`. What do you say we do this refactor at the same time we move it there?

salimbraksa replied:
@Gio2018 Sounds good to me 👍

salimbraksa commented:
@Gio2018 Great work covering the whole migration flow end-to-end. I've dropped a few comments sharing my thoughts on certain things but nothing is blocking. 🎉

Commit:
Fix trailing space violations

Commit:
Update migration view models, improve code legibility

Commit:
Update MigrationDependencyContainer, extract view controller factory, now passed to the navigation view controller

Commit:
Update MigrationNavigationController, use view controller factory to instantiate and push vcs to the stack

Commit:
Merge branch trunk into task/migration-notifications-premission

Gio2018 commented:
Thank you for the thorough review @salimbraksa !! 
I addressed the comments, change base to trunk and put it in ready for review, now ready for another pass!!

Commit:
Update JetpackWindowManager, simplify logic and make sure that the ui is always instantiated on the main thread

Commit:
Update MigrationFlowCoordinator, add logic to skip notifications permission screen if users have already decided

Commit:
Update MigrationNavigationController, always create and assign view controllers on the main thread

Commit:
Update migration view models, add notification permission logic and fix indentation

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Common/MigrationDependencyContainer.swift:
```diff
@@ -4,9 +4,29 @@ class MigrationDependencyContainer {
-                                      migrationStack: [.welcome: makeWelcomeViewController(),
-                                                       .notifications: makeNotificationsViewController(),
-                                                       .done: makeDoneViewController()])
+                                      factory: MigrationViewControllerFactory(coordinator: migrationCoordinator))
```
I like this, it's better than passing a closure 👍

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Common/Navigation/MigrationNavigationController.swift:
```diff
@@ -50,7 +50,9 @@ class MigrationNavigationController: UINavigationController {
-            self?.updateStack(for: step)
+            DispatchQueue.main.async {
+                self?.updateStack(for: step)
+            }
```
We can also use `.receive(on:)` if you prefer:
```swift
publisher
.receive(on: DispatchQueue.main)
.sink { value in
}
```

Source: https://developer.apple.com/documentation/combine/fail/receive(on:options:)

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Common/Navigation/MigrationFlowCoordinator.swift:
```diff
@@ -1,20 +1,30 @@
 /// Coordinator for the migration to jetpack flow
 final class MigrationFlowCoordinator: ObservableObject {
 
     @Published private(set) var currentStep = MigrationStep.welcome
```
Perhaps we should add a documentation here mentioning that `currentStep` values are published in a background thread.

I'm pretty sure future developers can easily know this by reading the `transitionToNextStep` implementation details but it would be great to make finding about this detail extra easier.

Gio2018 replied:
done!

salimbraksa commented:
@Gio2018 I've dropped some comments but LGTM! PR Approved 🎉 

_P.S: Don't forget to disable `shouldShowMigrationUI` flag before merging with `trunk`._

salimbraksa commented:
@Gio2018 The latest changes look good to me and I actually went ahead and approved the PR but then I discovered there is a tiny UI glitch where the app appears black for a second before the Welcome Screen shows up.

And I think the bug's root cause is:

```swift
init(coordinator: MigrationFlowCoordinator, factory: MigrationViewControllerFactory) {
   self.coordinator = coordinator
   self.factory = factory
   // `super.init(rootViewController:)` should be called for a smooth launch
   super.init(nibName: nil, bundle: nil)
   configure()
}
```

<img src="https://user-images.githubusercontent.com/9609223/200602995-0c528b72-4212-4cc2-9c8d-a63ed53532aa.png" width=400 />

Commit:
Update MigrationDependencyContainer, changed to struct since we don't need to hold a reference to it

Commit:
Update MigrationNavigationController, instantiate with rootViewController to avoid initial glitch

Gio2018 commented:
> there is a tiny UI glitch

good catch @salimbraksa ! I went ahead and changed the init to assign a root right away!

Commit:
Update MigrationViewControllerFactory, add method initialViewController()

Commit:
Update MigrationNavigationController, set initial root from factory and drop first state change, since it already corresponds to the same initial vc

Commit:
Update MigrationFlowCoordinator, add comments

salimbraksa commented:
@Gio2018 LGTM! 🚀

Commit:
Update JetpackWindowManager, small refactor

Commit:
Update JetpackWindowManager, small refactor

Commit:
Update JetpackWindowManager, disable migration UI
